File: gusah009/Java/[PGS]타겟넘버.java
class Solution {

  private static int TARGET;

  public int solution(int[] numbers, int target) {
    TARGET = target;
    int answer = dfs(numbers, 0, 0);

    return answer;
  }

  private int dfs(int[] numbers, int depth, int prevResult) {
    int result = 0;

    if (depth == numbers.length) {
      if (prevResult == TARGET) {
        return 1;
      }
      return 0;
    }

    result += dfs(numbers, depth + 1, prevResult + numbers[depth]);
    result += dfs(numbers, depth + 1, prevResult - numbers[depth]);

    return result;
  }
}

File: hall_of_fame/gusah009/[PGS]타겟넘버.cpp
#include <iostream>
#include <bits/stdc++.h>

using namespace std;

#define FOR(i, j) for(int i = 0; i < j; i++)

int N;
int numbers[21] = {0};
int target;
int answer = 0;

void input()
{
  cin >> N;
  FOR(n, N) {
    cin >> numbers[n];
  }
  cin >> target;
}

void DFS(int sum, int depth)
{
  if (depth == N - 1) {
    if (sum == target) {
      answer++;
    }
    return;
  }

  DFS(sum + numbers[depth + 1], depth + 1);
  DFS(sum - numbers[depth + 1], depth + 1);
  
}

void solve()
{
  DFS(numbers[0], 0);
  DFS(-numbers[0], 0);
}

void print()
{
  cout << answer << '\n';
}

int main()
{
  ios_base :: sync_with_stdio(false);
  cin.tie(NULL);
  cout.tie(NULL);
  input();
  solve();
  print();
}